[#16] 친구추가 요청 수락 기능 구현

msugo1 opened the pull request:
## FriendController

### `acceptFriendRequest` 메소드 추가

- 친구추가 수락에 대한 요청을 받을 컨트롤러 메소드

## FriendService

### `accpetFriendRequest` 메소드 추가

- 친구추가 수락 요청에 대한 비즈니스 로직을 수행할 메소드

-  DB에 등록된 상태가 `친구 요청을 받은 상태(RECEIVED)`와 일치하는 경우에만 요청을 Mapper로 전달

## FriendMapper

### `updateFriendRequest` 메소드 추가

- 비즈니스 계층으로부터 전달받은 친구 요청 수락에 대한 내용을 DB에 반영할 메소드

- `FriendMapper.xml` 에 위의 사항을 수행할 쿼리 추가

## 테스트

- 친구추가 요청 수락에 대한 로직을 테스트 할 테스트 작성

- 프스트맨을 이용한 테스트

Commit:
[#16] 친구추가 요청 수락 기능 구현

FriendController

* acceptFriendRequest 메소드 추가

- 친구추가 수락에 대한 요청을 받을 컨트롤러 메소드

FriendService

* accpetFriendRequest 메소드 추가

- 친구추가 수락 요청에 대한 비즈니스 로직을 수행할 메소드

: DB에 등록된 상태가 `친구 요청을 받은 상태(RECEIVED)`와 일치하는
경우에만요청을 Mapper로 전달

FriendMapper

* updateFriendRequest 메소드 추가

- 비즈니스 계층으로부터 전달받은 친구 요청 수락에 대한 내용을 DB에
  반영할 메소드

FriendMapper.xml

- 위의 사항을 수행할 쿼리 추가

테스트

- 친구추가 요청 수락에 대한 로직을 테스트 할 테스트 작성

- 프스트맨을 이용한 테스트

Commit:
[#17] 친구추가 요청 거절 기능 구현

FriendController

* rejectFriendRequest 메소드 추가

- 친구추가 거절에 대한 요청을 받을 컨트롤러 메소드

FriendService

* rejectFriendRequest 메소드 추가

- 친구추가 거절 요청에 대한 비즈니스 로직을 수행할 메소드

: DB에 등록된 상태가 `친구 요청을 받은 상태(RECEIVED)`와 일치하는
경우에만 요청을 Mapper로 전달

테스트

- 친구추가 요청 거절에 대한 로직을 테스트 할 테스트 작성

- 포스트맨을 이용한 테스트

Commit:
[#18] 친구 삭제 기능 구현

FriendController

* unfriendFriend 메소드 추가

- 친구삭제에 대한 요청을 받을 컨트롤러 메소드

FriendService

* unfriendFriend 메소드 추가

- 친구삭제 요청에 대한 비즈니스 로직을 수행할 메소드

: DB에 등록된 상태가 `친구 상태(FRIENDED)`인 경우에만 요청을 Mapper로 전달

테스트

- 친구 삭제 요청에 대한 로직을 테스트 할 테스트 작성

- 포스트맨을 이용한 테스트

Commit:
[#19] 로그인 체크 기능 구현

* HandlerInterceptor & `@LoginRequired`

- 인터셉터와 어노테이션을 활용해 현재 사용자가 로그인 했는지 먼저
  확인하는 기능을 구현

- ArgumentResolver와 `@CurrentUser`를 세션에 있는 아이디를 주입하기 전
  먼저 로그인 여부를 확인 하기 위함

* UnauthorizedAccessException

- 로그인하지 않은 사용자가 로그인이 필요한 기능에 접근 시 예외를
  던지는데 필요한 클래스

* ControllerExceptionHandler

- 위의 예외를 로그인 관련 ExceptionHandler 메소드에 매핑 & 문구 수정

- `friendRequestException` 메소드가 반환하는 예외 `title`이 좀 더
  범용적으로 활용될 수 있도록 변경

* WebConfig

- 위에서 작성한 HandlerInterceptor 등록

* FriendController, LanguageController, MyPageController

- 각 컨트롤러 메소드가 `@CurrentUser`을 파라미터로 주입 받고 있는 경우
  `@LoginRequired`를 붙여서 인터셉터가 동작하도록 설정

- FriendController에서 불필요하게 들어간 `@ResponseStatus(HttpStatus.OK)` 모두 삭제

msugo1 commented:
각 컨트롤러 메소드 위에 달려 있는 불필요한 `@ResponseStatus(HttpStatus.OK)`를 무의식적으로 붙였다는 것을 로그인 확인 기능 구현(#44  ) 중에 발견해서 모두 삭제 한 뒤 해당 브랜치에서 PR  날렸습니다. 앞으로 무의식적인 코드 작성을 좀 더 경계하도록 하겠습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/FriendController.java:
```diff
@@ -24,4 +24,10 @@ public void sendFriendRequest(@CurrentUser String userId, @PathVariable String t
         friendService.cancelFriendRequest(userId, targetId);
     }
+
+    @ResponseStatus(HttpStatus.OK)
```
기본이 200이므로 이것도 생략이 가능합니다~

f-lab-dev commented on src/main/java/me/soo/helloworld/service/FriendService.java:
```diff
@@ -52,13 +52,21 @@ private void isStatusValid(FriendStatus status) {
-        FriendStatus friendStatus = getFriendStatus(userId, targetId);
-
-        if (!friendStatus.equals(REQUESTED)) {
+    private void isStatusValid(FriendStatus currentStatus, FriendStatus targetStatus) {
```
1. `is`라는 말은 리턴값이 boolean일때 붙입니다~ `validStatus`가 어떨까요?
2. 보통 private 메소드는 제일 아래에 위치시킵니다~

msugo1 replied:
1. 넵. 메소드가 해당 클래스의 `행동`을 의미하니까 valid의 동사 형태인 `validate`를 사용하였는데 괜찮은지 모르겠네요.
2. private 메소드를 맨 아래로 보냈습니다.

**질문**
- 현재 `validateStatus` 메소드가 두 개가 있는데요. 하나는 `status` 만 집어 넣어서 동작하는 메소드인데 친구 요청시에만 사용가능한 대신에 케이스별로 나눠서 예외 메세지를 좀 더 자세히 전달할 수 있다는 장점이 있긴 한데, 생각할 수록 두 번째 `validateStatus`아래 메소드로 대체가 가능하긴 할 것 같더라고요. 다만 대체해버리면 예외 메세지가 `잘못된 접근입니다.`로 너무 단순해 질 것 같은데 `자세한 예외메세지 vs 약간 더 깔끔할 코드` 중에서는 그냥 자세한 예외 메세지를 선택하는 것이 나중에 유지보수 등에 더 도움이 될까요?

f-lab-dev replied:
`validateStatusDetail` 같이 역할을 나눠서 필요에 맞게 사용하는건 어떨까요?

msugo1 replied:
일단 친구요청을 보내는 행동 자체가 수락/거절/취소 등 다른 모든 행동의 시발점이니 좀 더 자세한 예외메시지를 보내주는게 맞는 것 같아요. 다른 요청들은 친구요청처럼 디테일하지 않아도 무엇이 잘못되었는지 문제를 인식하는데 큰 지장이 없을 것 같기도 하구요. 

그럼 해당 메소드 이름을  `validateStatusDetail`로 변경해놓고 계속 리팩토링/다른 내용 구현 고민하면서 좀 더 자세한 예외를 던질 필요가 있으면 비슷한 유형의 다른 메소드를 만들거나 혹은 해당 메소드 내용을 수정해나가는 선택은 어떨까요?

계속 고민하다가 이중 스위치문을 사용하면 하나의 내용으로 구현가능하다는 것을 깨닫긴 했는데 쓸데없이 길어지고 가독성도 나빠져 이해하기도 쉽지 않더라구요.

Commit:
[#16] 피드백 반영

- 메소드 이름 변경 (isStatusValid -> validateStatus)

- private 메소드 위치를 클래스의 제일 아래로 보내기

Commit:
[#17] 피드백 반영

- 불필요한 `@ResponseStatus` 제거 (HttpStatus.OK)
- `Friends` 테이블의 Schema를 담은 `friends.sql` 파일 추가

Commit:
[#18] 피드백 반영

- 불필요한 `@ResponseStatus` 제거

Commit:
[#19] 피드백 반영

1. LoginInterceptor의 preHandle

- 메소드 내에서 현재 로그인 중인 사용자의 아이디를 세션으로 받을 때
  캡슐화가 되어 있는 `LoginService`의 `getCurrentUserId()`메소드를
  사용하는 것으로 변경

2. ExceptionHandler내의 로그인 실패/인증 예외 분리

- 합쳐져 있던 로그인 실패와 인증 관련 예외를 분리

msugo1 committed:
Merge pull request #40 from f-lab-edu/feature/17

[#17] 친구추가 요청 거절 기능 구현

msugo1 committed:
Merge pull request #41 from f-lab-edu/feature/18

[#18] 친구 삭제 기능 구현

msugo1 committed:
Merge pull request #45 from f-lab-edu/feature/19

[#19] 로그인 체크 기능 구현

f-lab-dev commented on src/main/resources/sql/friends.sql:
```diff
@@ -0,0 +1,11 @@
+);
+
+create index friends_userId_index
+    on friends (userId);
```
`friendId`는 일부러 인덱스를 걸어주시지 않은걸까요?

msugo1 replied:
인덱스 정하는 기준이 검색 조건에 자주 쓰이는 경우라고 하더라구요. 그래서 생각해봤을 때 친구를 검색하는 기준이 친구 상태의 `status`, `userId`, `id`로 `friendId`는 굳이 쓰이지 않을 것 같아서 인덱싱에서 제외하였습니다.

msugo1 committed:
Merge branch 'develop' into feature/16

Commit:
[#16] 피드백 반영

* `validateStatus` ->  `validateStatusDetail`로 메소드 이름 변경

: 친구 요청을 보낼 때 더 자세한 예외 메세지를 전달하기 위해서 일반
validateStatus 메소드와 구분

Commit:
Merge branch origin 'feature/16' into feature/16
